tests/main/snap-session-agent-service-control/task.yaml

```diff
@@ -0,0 +1,88 @@
+    # be the case on distributions where presets have been used to
+    # disable it.
+    if [ ! -L /usr/lib/systemd/user/sockets.target.wants/snapd.session-agent.socket ] &&
+            ! systemctl --user --global is-enabled snapd.session-agent.socket; then
+        systemctl --user --global enable snapd.session-agent.socket
+        touch agent-was-enabled
+    fi
+    mkdir -p "$USER_RUNTIME_DIR"
+    chmod u=rwX,go= "$USER_RUNTIME_DIR"
+    chown test:test "$USER_RUNTIME_DIR"
+    systemctl start "user@${TEST_UID}.service"
+
+    # ensure curl is available (needed for e.g. core18)
+    if ! command -v curl; then
+        snap install --devmode --edge test-snapd-curl
+        snap alias test-snapd-curl.curl curl
+    fi
+
+restore: |
+    if snap list test-snapd-curl; then
+        snap remove test-snapd-curl
+    fi
+    systemctl stop "user@${TEST_UID}.service"
+    rm -rf "${USER_RUNTIME_DIR:?}"/* "${USER_RUNTIME_DIR:?}"/.[!.]*
+    if [ -f agent-was-enabled ]; then
+        systemctl --user --global disable snapd.session-agent.socket
+        rm agent-was-enabled
+    fi
+    rm -f /etc/systemd/user/snap.test-service.service
+
+execute: |
+    systemctl_user() {
+        su -l -c "XDG_RUNTIME_DIR=\"${USER_RUNTIME_DIR}\" systemctl --user $*" test
+    }
+
+    echo "Create a user mode service"
+    cat << \EOF > /etc/systemd/user/snap.test-service.service
+    [Unit]
+    Description=test service
+    [Service]
+    Type=simple
+    ExecStart=/bin/sleep 1d
+    EOF
+
+    echo "The session agent can reload the user mode systemd instance"
+    curl --unix-socket "${USER_RUNTIME_DIR}/snapd-session-agent.socket" \
+        -D- -X POST -H "Content-Type: application/json" \
+        -d '{"action": "daemon-reload"}' \
+        http://localhost/v1/services | MATCH "HTTP/1.1 200 OK"
+
+    echo "The service is now visible but not active"
+    systemctl_user show --property=LoadState snap.test-service.service |
+        MATCH LoadState=loaded
+    systemctl_user show --property=ActiveState snap.test-service.service |
+        MATCH ActiveState=inactive
+
+    echo "The session agent can start user mode services"
+    curl --unix-socket "${USER_RUNTIME_DIR}/snapd-session-agent.socket" \
+        -D- -X POST -H "Content-Type: application/json" \
+        -d '{"action": "start", "services": ["snap.test-service.service"]}' \
+        http://localhost/v1/services | MATCH "HTTP/1.1 200 OK"
+    systemctl_user show --property=ActiveState snap.test-service.service |
+        MATCH ActiveState=active
+
+    echo "The session agent can stop user mode services"
+    curl --unix-socket "${USER_RUNTIME_DIR}/snapd-session-agent.socket" \
+        -D- -X POST -H "Content-Type: application/json" \
+        -d '{"action": "stop", "services": ["snap.test-service.service"]}' \
+        http://localhost/v1/services | MATCH "HTTP/1.1 200 OK"
+    systemctl_user show --property=ActiveState snap.test-service.service |
+        MATCH ActiveState=inactive
```

Comment:
I've created PR #7555 with a spread test demonstrating this for strict confinement systems.  I agree that we might want to add further security checks for partial confinement systems, but it probably wouldn't have a meaningful impact on security even with this new session agent API.

If the app can connect to the session bus (either because there is no file level AppArmor protection, or because it has connected one of the interfaces that allows it), then it will be able to control the user instance of systemd directly.  At that point, it will be able to do everything this API allows and more.

Reply:
Thank you for PR 7555. I left my review here: https://github.com/snapcore/snapd/pull/7555#pullrequestreview-334233943 (approved)